Add PubSub and BigTable support for Central hosted network controllers

glimberg opened the pull request:
This is a big one that goes along with other internal upcoming changes.

* Refactors existing Redis & Postgres NOTIFY message passing systems into the `NotificationListener` interface, and adds a 3rd method for message passing via GCP PubSub.  Which one to use is configurable by local.conf controller settings
* Adds GCP BigTable as a 3rd option for writing member status updates along side Postgres and Redis.  Also configurable via local.conf
* Set up CMake for building Central controllers and extended for the regular agent build as well (optional use threre. existing build methods still exist).
* `scripts/bootstrap-deps.sh` for bootstrapping external dependencies (`scripts/bootstrap-deps.ps1` on Windows)
* CMake for external dependency management instead of dumping everything in `ext/`.  This is likely more useful for Central controller builds than end user builds for the time being. 
* Controller log formatting standardization

glimberg committed:
reorganize rustybits into a single library with smeeclient behind the ztcontroller feature flag

glimberg committed:
fix dependencies when temporal isn't needed

glimberg committed:
gcloud-pubsub only if ztcontroller is flagged

glimberg committed:
fix macos build

glimberg committed:
fix windows bui ld

glimberg committed:
tokio is needed by both temporal & gcloud pubsub, so make just one instance for the whole library, add init/shutdown functions for it exposed to C

glimberg committed:
cleanup warnings

glimberg committed:
WIP: pubsub wrapper in Rust

glimberg committed:
plumb through callbacks

glimberg committed:
function naming cleanup.

match other things already written

glimberg committed:
updates & tests.

Tests currently need to be run with --test-threads=1.  Seems like the instances of the pubsub emulator stomp on each other without that

glimberg committed:
expose change_handler to C via FFI

glimberg committed:
remove ztcontroller from default feature list

glimberg committed:
feature cleanup

glimberg committed:
fix build script

glimberg committed:
fix calling into async functions from non-async via the FFI

glimberg committed:
revert default feature list once again

glimberg committed:
more calling async from non-async changes

glimberg committed:
Add C++ wrapper around pubsub listeners

glimberg committed:
rename classes

glimberg committed:
Refactor Redis & Posgres notification listeners into listener subclass in new CentralDB class

This allows us to interchangeably use different listeners (pgsql, redis, pubsub) depending on configuration values passed into the constructor.

glimberg committed:
make whether SSO is enabled a switchable config value

glimberg committed:
rework protobuf messages

PubSub allows us to do schema validation, however it only allows one top level message at a time.  Move other sub-message declarations under the main message declaration so that we can enable schema validation in the pubsub stream directly

glimberg committed:
Update central controller build to use CMake + conda

Muuuuch easier to use external dependencies now

Also tried out conan and vcpkg.   Ran into dependency issues when solving for packages to install with conan.

vcpkg is just obtuse as all hell to install and not easy to integrate

glimberg committed:
spacing

glimberg committed:
clang format change (not applied project-wide yet)

glimberg committed:
add a StatusWriter class hierarchy for writing member status updates

* Postgres Direct
* Redis
* BigTable

glimberg committed:
wire up status writers

glimberg committed:
add sleep at end of online notificaiton loop

glimberg committed:
Merge branch 'dev' into gl/ctl-pubsub

glimberg committed:
PubSub to gcloud-cpp library for simplicity's sake

glimberg committed:
rework the pubsub listener again so the subscription is set up on its own thread

glimberg committed:
remove rust-based pubsub

glimberg committed:
add def for new style controllers

glimberg committed:
Merge branch 'adam/1.16' into gl/ctl-pubsub

glimberg committed:
Create the pubsub topic if it doesn't exist

glimberg committed:
Wiring through initialization of the CentralDB version of the controller

Still need to do the actual configuration from local.conf

glimberg committed:
wire up configuration.

New config block in local.conf for controllers:

```
{
  "settings": {
    ...standard zt1 local.conf settings...
  },
  "controller": {
    "listenMode": (pgsql|redis|pubsub),
    "statusMode": (pgsql|redis|bigtable),
    "redis": {
      "hostname": ...,
      "port": 6379,
      "clusterMode": true
    },
    "pubsub": {
      "project_id": <gcp-project-id>
    },
    "bigtable": {
      "project_id": <gcp-project-id>,
      "instance_id": <bigtable-instance-id>,
      "table_id": <bigtable-table-id>
    }
  }
}
```

glimberg committed:
update central controller readme with config block

glimberg committed:
remove otel warning

glimberg committed:
setup github action for building

glimberg committed:
runs-on matrix

glimberg committed:
skip caching until this all works

glimberg committed:
trying to get this working

glimberg committed:
test

glimberg committed:
no linux/arm64 build available for google-cloud-cpp 2.39 :/.  Downgrade to 2.38

glimberg committed:
dependency juggling

glimberg committed:
fix cmake options

glimberg committed:
more conda

glimberg committed:
more conda

glimberg committed:
skip cmake action

glimberg committed:
cmake --build args

glimberg committed:
fix arm64 build flag

glimberg committed:
fix deps for selftest

glimberg committed:
set ZT_AES_NEON for arm64

glimberg committed:
cache conda environment

glimberg committed:
cache conda environment

glimberg committed:
fix run arg

glimberg committed:
bump

glimberg committed:
put git branch ref into cache key name

glimberg committed:
only update conda env if cache was not hit

glimberg committed:
let's build a docker image

glimberg committed:
fix short hash

glimberg committed:
fix multi-arch manifest creation

glimberg committed:
update ld.so.conf for paths to libraries

glimberg committed:
new modified startup script for central controller docker images with support for configuring bigtable, pubsub, redis

glimberg committed:
no trailing comma on bigtable block

glimberg committed:
text alignment

glimberg committed:
add jemalloc

glimberg committed:
change tag to match existing

glimberg committed:
need a conda init here first apparently

glimberg committed:
stray backtick

glimberg committed:
also have to add postgres to the container for pg_isready

glimberg committed:
no absolute path for pg_isready

glimberg committed:
ok back to full path

glimberg committed:
run process in conda env directly from docker

glimberg committed:
debug output of main script

glimberg committed:
.

glimberg committed:
fix case for redis: null

glimberg committed:
and instead of or in postgres check

glimberg committed:
debug printf

glimberg committed:
get a running controller in docker

glimberg committed:
debugging runtime link issues on arm64

glimberg committed:
try to make this work on both x64 & arm64

glimberg committed:
will this make arm64 containers work?

glimberg committed:
add conda builder dockerfile

glimberg committed:
trying to build in a docker container

glimberg committed:
build in containers

glimberg committed:
move up docker auth

glimberg committed:
move up the gcp auth step

glimberg committed:
cleanup controller github action

glimberg committed:
enable otel exporter

glimberg committed:
fix otel exporting

glimberg committed:
query fix

glimberg committed:
another query fix

glimberg committed:
another query fix

glimberg committed:
another query fix

glimberg committed:
fix

glimberg committed:
fix

glimberg committed:
another fix

glimberg committed:
add a ChangeSource enum to the change protobuf messages

glimberg committed:
Add `frontend` column to networks & network_memberships tables so the controller knows which system owns the networks & members

glimberg committed:
wire up pubsub outgoing status changes from controller -> CV2

glimberg committed:
whoops. mesed up the migrate naming convention

glimberg committed:
misc bugfixes

glimberg committed:
apply patch to libnatpmp repo so we can compile on macOS with CMake

glimberg committed:
compiler on mac doesn't like this.  not really needed though

glimberg committed:
fix compile & link on macOS

glimberg committed:
fix stack corruption in DB::get

glimberg committed:
fix typo in cmake file

glimberg committed:
more fixes

glimberg committed:
alloc meatadata with new so pubsub can delete it

glimberg committed:
delete pubsub subscription on shutdown

glimberg committed:
append 8 random hex chars to the subscription id

glimberg committed:
forgot a ;

glimberg committed:
pass version info to status writer

glimberg committed:
back to bulk apply for status updates in BigTable

glimberg committed:
only update os, architecture and client version if they've changed (or don't exist)

glimberg committed:
removing some log output

glimberg committed:
wire up pubsub notifications from controller to frontend

glimberg committed:
clean up deprecation warnings in pqxx

glimberg committed:
fix version string

glimberg committed:
add `use_for_new_networks` column in controller table

glimberg committed:
add tables for the controller load log and sso nonces & expiry

glimberg committed:
fix migration

glimberg committed:
make pubsub topics configurable

Also for pubsub trips back to CV1/CV2, rather than having 2 queues for each service (networks & members), theres now only a single queue for each change type, and `frontend = (cv1|cv2)` attribute is set on the message for filtering.

glimberg committed:
try making the controller attach to an existing pubsub subscription if it already exists

glimberg committed:
only create the subscription if pubsub emulator is being used.

glimberg committed:
set objtype on messages received via PubSub

glimberg committed:
fix subscription creation

glimberg committed:
query fix for new networks

glimberg committed:
json field fix

glimberg committed:
more logging  for errors on pubsub listener

glimberg committed:
stray ` got in there

glimberg committed:
fix attempted null -> string conversion

glimberg committed:
debug output from parsing pubsub messages

glimberg committed:
debug output

glimberg committed:
more logging

glimberg committed:
more logging

glimberg committed:
more debug output

glimberg committed:
make the default an empty array here

glimberg committed:
fix tags/capabilities?

glimberg committed:
think this is where things are crashing, but I'm not sure why yet.  Comment it out to make sure

glimberg committed:
throwing stuff at the wall cuz I can't actually debug this right now

glimberg committed:
pubsub writer should work now

glimberg committed:
one last json parsing fix

glimberg committed:
one last json parsing fix.  but really this time

glimberg committed:
cleaning up some gross JSON code

glimberg committed:
add explicit nack if there's an error processing a pubsub message

glimberg committed:
Fixing more JSON issues

glimberg committed:
no need for these timeouts

glimberg committed:
debug output for IP addressing & fixing order of operations in a couple of places.  Only send notification of a change to pubsub after it's been written to the DB

glimberg committed:
fix assign mode json

glimberg committed:
Merge branch 'dev' into gl/ctl-pubsub

glimberg committed:
dont NACK failed messages.  That gets us into a loop of failing the same messages over and over again

glimberg committed:
robustify the handling of tags/capabilities/rules which are strings of encoded json

glimberg committed:
more robustifying incoming messages

glimberg committed:
compiler fix

glimberg committed:
more variations on checking json

glimberg committed:
these should be empty json arrays, not just "[]".

Actually these *should* be structs, not filling out json.  Adding this to the tech debt backlog

glimberg committed:
missed one

glimberg committed:
ensure change source is controller if otherwise unset

glimberg committed:
Added some more logging to the ctl around deauths

glimberg committed:
Set ordering key on pubsub message publishing.

glimberg committed:
Enable cross-service propagation of otel trace metadata

joseph-henry commented:
@glimberg I'm only just noticing this. Do you still want it merged in? I see it is still a draft.

glimberg commented:
Eventually yes.  Just making sure everything is stable over in the new environment before doing so

glimberg committed:
Merge branch 'dev' into gl/ctl-pubsub

glimberg committed:
controller build readme update

glimberg commented:
@joseph-henry ready for your review.  I know this is a huge one. Don't hesitate to reach out if you have any questions

glimberg commented:
OK not sure why the github action is failing now but I'll look at it tomorrow

glimberg committed:
set --provenance false on docker build to try and fix docker image creation issue

glimberg committed:
Merge branch 'dev' into gl/ctl-pusub

glimberg committed:
Remove extra verbose logging from controller

glimberg committed:
disable peer metrics in Central controller

glimberg committed:
WIP: Update sso info retrieval method

glimberg committed:
add a linked_id column to the oidc_config table.

Required to get the list of configs  for a particular org that the controller has.  Named it `linked_id` rather than `org_id` since we don't know what it will be linked to in CV2

glimberg committed:
fix db migrations

glimberg committed:
drop index

glimberg committed:
woops.  out of order here

glimberg committed:
set network member frontend based on the network its a member of

glimberg committed:
add SSO nonce/auth PubSub messaging with frontend routing

Publish CTL_NONCE_UPDATE to PubSub when nonces are created or reused in
getSSOAuthInfo(), with the network's frontend as a message attribute so
only the correct CV frontend receives it. Listen for ZT1_AUTH_UPDATE
messages and update sso_expiry.authentication_expiry_time accordingly,
with a network existence check before applying.

- Add sso_send_topic/sso_recv_topic to PubSubConfig
- Add PubSubWriter::publishSSONonceUpdate() with frontend param
- Add PubSubSSOListener class for inbound auth updates
- Rename CV1_AUTH_UPDATE to ZT1_AUTH_UPDATE in sso.proto
- Fix pre-existing connection pool leak in getSSOAuthInfo() catch block

glimberg committed:
add assigned_central_version column to controllers_ctl

Allow controllers to advertise which central version (cv1, cv2, or all)
they are assigned to handle via a new configurable field. The value is
persisted to the database on each heartbeat and validated at startup
against the DB CHECK constraint.

glimberg committed:
Undo change to old migration that shouldn't have been made

glimberg committed:
another fix

glimberg committed:
plumb through config changes for sso pubsub

glimberg committed:
configure assigned central version in startup script

glimberg committed:
print a message when the SSO PSK is configured

glimberg committed:
temporary logging

glimberg committed:
update settings to enable SSO networks

glimberg committed:
sso query fix in controller

glimberg committed:
Skip redundant nonce sending with an expiry time of 0

glimberg committed:
Remove smee from CentralDB.

Now handled in CV1 on new member join via pubsub integration when a new member comes through

glimberg committed:
Added a little bit more logging for the node checkin/bigtable write process

glimberg committed:
periodic queue size logging, and fix some db connection leaks

cspiegel commented on nonfree/controller/RedisStatusWriter.cpp:
```diff
@@ -0,0 +1,123 @@
+
+	std::set<std::string> networksUpdated;
+	uint64_t updateCount = 0;
+	for (const auto& entry : _pending) {
```
`_pending` here should be `toWrite`.

cspiegel commented on nonfree/controller/RedisStatusWriter.cpp:
```diff
@@ -0,0 +1,123 @@
+
+namespace ZeroTier {
+
+RedisStatusWriter::RedisStatusWriter(std::shared_ptr<sw::redis::Redis> redis, std::string controller_id)
+	: _redis(redis)
+	, _mode(REDIS_MODE_STANDALONE)
+{
+}
+
+RedisStatusWriter::RedisStatusWriter(std::shared_ptr<sw::redis::RedisCluster> cluster, std::string controller_id)
+	: _cluster(cluster)
+	, _mode(REDIS_MODE_CLUSTER)
+{
+}
```
These constructors don't set `_controller_id`.

cspiegel commented on nonfree/controller/EmbeddedNetworkController.cpp:
```diff
@@ -594,6 +646,47 @@ void EmbeddedNetworkController::init(const Identity& signingId, Sender* sender)
+		throw std::runtime_error("controller config required");
+	}
+
+	if (_path.length() > 9 && (_path.substr(0, 9) != "postgres:")) {
```
This test isn't quite complete enough: it allows shorter, non-postgres strings through. You can just flip the size check:

```if (_path.length() < 10 || (_path.substr(0, 9) != "postgres:"))```

cspiegel commented on nonfree/controller/CentralDB.cpp:
```diff
@@ -0,0 +1,1807 @@
+			}
+			else {
+				throw std::runtime_error("CentralDB: Redis listener mode selected but no Redis configuration provided");
+			}
+		case LISTENER_MODE_PUBSUB:
```
Missing `break` here.

cspiegel commented on nonfree/controller/CentralDB.cpp:
```diff
@@ -0,0 +1,1807 @@
+								nlohmann::json oldMember;
+								nlohmann::json newMember = config;
+								if (! isNewMember) {
+									oldMember = _getNetworkMember(w, networkId, memberId);
```
Above was a `w.commit()`, after which any  queries on `w` _should_ be raising exceptions, since the transaction is complete. I'm wondering if this is a code path that has never been hit in testing (i.e. pubsub, null/controller change source, and it's not a new member).

glimberg replied:
this one should actually be fine.  Only one branch of this gets hit per function call based on objtype, so a commit() in ea

cspiegel commented on nonfree/controller/CentralDB.cpp:
```diff
@@ -0,0 +1,1807 @@
+																   frontend);
+						}
+
+						w.commit();
```
Similar to elsewhere, this commit will be "invalidating" the handle (since the transaction is complete). That would mean a call such as on line 541 shouldn't properly work. Is this a path that isn't occurring or is my understanding of libpqxx's transaction handling obsolete?

cspiegel commented on nonfree/controller/CentralDB.cpp:
```diff
@@ -0,0 +1,1807 @@
+{
+	fprintf(stderr, "%s: commitThread start\n", _myAddressStr.c_str());
+	_queueItem qitem;
+	while (_commitQueue.get(qitem) & (_run == 1)) {
```
This should be `&&` instead of `&`.

cspiegel commented on nonfree/controller/CentralDB.cpp:
```diff
@@ -0,0 +1,1807 @@
+		AuthInfo info;
+		info.enabled = true;
+
+		// if (memberId == "a10dccea52" && networkId == "8056c2e21c24673d") {
+		//	fprintf(stderr, "invalid authinfo for grant's machine\n");
+		//	info.version=1;
+		//	return info;
+		// }
+		//  fprintf(stderr, "CentralDB::updateMemberOnLoad: %s-%s\n", networkId.c_str(), memberId.c_str());
```
Leftover commented code.

cspiegel commented on nonfree/controller/CtlUtil.cpp:
```diff
@@ -63,6 +80,118 @@ std::string url_encode(const std::string& value)
+	}
+}
+
+// void create_bigtable_table(std::string project_id, std::string instance_id)
+// {
+// 	auto bigtableAdminClient =
+// 		bigtable_admin::BigtableTableAdminClient(bigtable_admin::MakeBigtableTableAdminConnection());
+
+// 	std::string table_id = "member_status";
+// 	std::string table_name = "projects/" + project_id + "/instances/" + instance_id + "/tables/" + table_id;
+
+// 	// Check if the table exists
+// 	auto table = bigtableAdminClient.GetTable(table_name);
+// 	if (! table.ok()) {
+// 		if (table.status().code() == google::cloud::StatusCode::kNotFound) {
+// 			google::bigtable::admin::v2::Table table_config;
+// 			table_config.set_name(table_id);
+// 			auto families = table_config.mutable_column_families();
+// 			// Define column families
+// 			// Column family "node_info" with max 1 version
+// 			// google::bigtable::admin::v2::ColumnFamily* node_info = table_config.add_column_families();
+// 			// Column family "check_in" with max 1 version
+
+// 			auto create_result = bigtableAdminClient.CreateTable(
+// 				"projects/" + project_id + "/instances/" + instance_id, table_id, table_config);
+
+// 			if (! create_result.ok()) {
+// 				fprintf(
+// 					stderr, "Failed to create Bigtable table member_status: %s\n",
+// 					create_result.status().message().c_str());
+// 				throw std::runtime_error("Failed to create Bigtable table");
+// 			}
+// 			fprintf(stderr, "Created Bigtable table: member_status\n");
+// 		}
+// 		else {
+// 			fprintf(stderr, "Failed to get Bigtable table member_status: %s\n", table.status().message().c_str());
+// 			throw std::runtime_error("Failed to get Bigtable table");
+// 		}
+// 	}
+// }
```
Is this leftover code, or intended to stay as documetation?

glimberg committed:
fix poison pill blocking proccessing

glimberg committed:
Fix silent PubSub message loss in controller subscription loop

The 10-second session.cancel() loop raced with in-flight acks — when
cancel fired while the GCP client was processing messages, acks were
lost before reaching the server. With message ordering enabled, an
unacked message blocks all subsequent messages on that ordering key,
causing silent stalls with no error output.

Two fixes:
- Replace the cancel/reconnect timer with a blocking session.get(),
  storing the session future so the destructor can cancel on shutdown.
- Always ack messages even when onNotification fails — permanent errors
  (bad protobuf, missing fields) will never succeed on retry and would
  otherwise poison the ordering key indefinitely.

glimberg committed:
logging

glimberg committed:
make sure subscribe pulls stay running

glimberg committed:
change sso topic names for clarity

glimberg committed:
addressing pr comments

glimberg committed:
fix a typo in some json handling that could be affecting an issue I'm tracking down

glimberg committed:
Bugfix pass

ssoConfig checked wrong variable
DNS missing-branch wrote to input not output
SSO fields renamed to sso* and nested under ssoConfig (across CentralDB, DB defaults, migration_models)
Member identity populated into the right field; address = memberId
noAutoAssignIPs → noAutoAssignIps casing
Member version columns now SELECTed + populated (both sites)
removeTraceLevel typo in DB.cpp defaults
8. eraseMember uninitialized tmp2 — member deletes now actually delete
9. Connection leak on continue in both commit branches
10. Unsafe (uint64_t)config["revision"] cast → OSUtils::jsonInt
11. target = "NULL" literal text → std::optional<std::string> for real SQL NULL
~13 unsafe direct json→string casts → OSUtils::jsonString; dead vars removed from save(); change_source blocks use is_string() checks.
Heartbeat zadd literal "controllerId" → variable
Network revision double-increment removed
PubSub notifier now sees actual old state (DB query moved before INSERT in both branches)
_delete_network reads config["id"] instead of config["nwid"]
ip_assignments empty-string guards added at both sites (also closes G)
ensure last_modified is set on every write to networks & members

glimberg committed:
Fix json key inconsistency in pubsub writer

glimberg committed:
align JSON keys between CentralDB and PubSub listener/writer

- Default enableBroadcast to true on load (initializeNetworks and _getNetwork) to match DB::initNetwork. Networks whose JSONB lacks the field now get broadcast on instead of off, which is the documented ZeroTier default.

- Stop overwriting networks_ctl.frontend on every UPDATE. It should remain whatever was set at INSERT. The early-exit guard already prevents cross-frontend writes from applying, so dropping the SET preserves correctness. Members are unaffected — they source frontend from the network row, not the change payload.

- Add enableBroadcast, creationTime, and revision to PubSubListener's network toJson. The protobuf carries them, but the listener was dropping them on inbound, which caused revision to reset on every PubSub-delivered network update.

- Rename member version keys to vMajor/vMinor/vRev/vProto in both PubSubListener (incoming) and PubSubWriter (outgoing). CentralDB, EmbeddedNetworkController, and the rest of the codebase already use these names; the listener/writer were the only sites using versionMajor/Minor/Rev/Protocol, so member version info was being silently dropped through PubSub in both directions.

- Pass newMember by const reference in ControllerChangeNotifier::notifyMemberChange to match notifyNetworkChange. Avoids a full JSON copy on every member notify.

- Remove duplicate nlohmann::json::parse call in RedisNetworkListener.

glimberg committed:
update network.proto in line with CV1 and CV2

glimberg committed:
fix for missing field in json

glimberg committed:
ensure controller doesn't change assigned frontend

glimberg committed:
handle nulls for remote trace target

glimberg committed:
remove noisy log

glimberg committed:
Merge branch 'dev' into gl/ctl-pubsub

glimberg committed:
cargo fix

glimberg committed:
This workflow does need to be removed from this branch.

once finally merged it'll be irrelevant.

glimberg committed:
More sso error logging & removing a few logs as well

glimberg committed:
Add handling sso networks for both CV1 & CV2 simultaneously

glimberg committed:
perf(controller): drop per-row BigTable reads in status writer

writePending() did a synchronous ReadRow per queued node to avoid
rewriting unchanged os/arch/version. With thousands of check-ins per
cycle that's thousands of sequential round-trips every 10s, and the read
was nearly pointless since last_seen and the IP cell are written every
cycle anyway.

Replace it with an in-process cache of the last-written node_info per
row (keyed by a 64-bit hash of the row key). node_info is emitted only
when changed, new, or past its refresh window; check_in is still written
every cycle. Failed rows are dropped from the cache so they're rewritten
next cycle, and stale entries are evicted (24h TTL). Net result: one
BulkApply per cycle, no read RPCs.

glimberg committed:
better closed connection handling to postgres

glimberg committed:
Better data handling if a db write fails

glimberg committed:
better logging of commit thread ticks

shows whats happening on each

glimberg committed:
remove db check from onlineNotificationThread

glimberg committed:
fix pubsub flood on ctl startup

glimberg committed:
Repo cleanup.

* Remove now dead separate CV1/CV2 implementations.  Consolidated into
  CentralDB.
* Remove smeeclient.  No longer needed.

glimberg committed:
main controller cmake file cleanup

glimberg committed:
bring CMake in line with make-linux/mac/bsd

glimberg committed:
Fix PIC/PIE flags

glimberg committed:
More linux/mac cmake cleanup

* Brewfile for homebrew dependencies on macos
* Pull & build google-cloud-cpp/otel via bootstrap script rather than
  keeping it all in the ext/ subfolder.
* Update controller CI build with new steps

glimberg committed:
build fix

glimberg committed:
Set thinkgs up for Windows in CMake

Also fixed a ZT_NONFREE issue

glimberg committed:
update to use makefiles by default on *ix

glimberg committed:
fix linking via CMake on Windows

glimberg committed:
fix incompatible options on crypto files in windows

glimberg committed:
fix deps so that architecture is taken into account

glimberg committed:
Bump cpp-httplib so it supports win32

glimberg committed:
more build updates.

removing openssl dependency (except on Linux where it's still necessary)
upgrade cpp-httplib
new cmake-presets

glimberg committed:
set vcpagkage baseline and remove comment with text windows doesn't like

glimberg committed:
Addressing comments

glimberg committed:
bugfixes

glimberg committed:
Pubsub devlivery reliability updates

glimberg committed:
Better connection handling for postgres status writer

glimberg committed:
status update batch delivery reliablilty

glimberg committed:
Concurrency hygene updates

glimberg committed:
Correcting a few things

glimberg committed:
fix printf, connection pool smart pointer semantics, wait for ready spin
lock -> notification, and double notify on erase

glimberg committed:
Conneciton pool update to a timed wait on connection pool exhaustion

More use of scoped RAII guard for connections where appropriate.

glimberg committed:
enable metrics by default for the central controller

glimberg committed:
use the checkin timestamps for the cell timestamps in BigTable

glimberg committed:
Update build.sh with build instructions

Also add presets for the BSDs

glimberg committed:
Ensure everything in nonfree/controller has the proper header comment
pointing to LICENSE.TXT

glimberg committed:
standardize log formatting across the controller

glimberg committed:
update documentation for ConnectionPool

glimberg committed:
bump actions to get rid of node warnings

glimberg committed:
Fix the sso template for errors returned by IdP

Template was looking for {{networkId}} but that hadn't been populated
yet in the case where the IdP returned an error. This caused the
template parsing to crash and return its own error, disguising the error
from the IdP

glimberg committed:
Fix a potential crash on central controller shutdown

glimberg committed:
Add a SIGSEGV handler to print a stack trace for central controllers

Only enabled in the central controller builds.  Trying to find more info
about some mystery controller restarts in prod

joseph-henry commented:
Working notes:
 - It looks like the Smeeclient was removed.
 - The non-CMake build process for ZT_CONTROLLER=1 is now broken. I propose we remove it entirely.

glimberg commented:
yes, smeeclient was removed as it's no longer needed for notifications of new members joining.  Thanks to pubsub, that is now handled in CV1/CV2 instead of the controller.

Will remove all the ZT_CONTROLLER=1 stuff from the makefile

glimberg committed:
Remove ZT_CONTROLLER=1 from makefile.  No longer used to build central
controllers

joseph-henry commented on nonfree/controller/ConnectionPool.hpp:
```diff
@@ -105,49 +172,48 @@ template <class T> class ConnectionPool {
-				// Nothing available
+			// Everything is in active use. Wait for an unborrow() to free a slot, then
+			// retry. Give up (throw, as before) once the deadline passes.
+			if (m_cond.wait_until(l, deadline) == std::cv_status::timeout) {
```
I assume this delay is to act as a rate limit to prevent something from getting spammed? Do we usually see a lot of bursty attempts here?

joseph-henry commented:
Did an end to end test (root, controller, two nodes) of this branch using artifacts produced by the CMake build process and the two nodes were able to ping each other so it looks like things still work.

Test output:

```
[2026-06-29 18:48:27] === ZeroTier Infrastructure Starting ===
[2026-06-29 18:48:27] === Regenerating root identity, controller identity, and signed planet ===
[2026-06-29 18:48:27] Generated root=74bea25db9 controller=5946b73234
wrote 0000000008eac90a.moon (signed world with timestamp 1782758907554)
[2026-06-29 18:48:27] Regen complete: root=74bea25db9 controller=5946b73234 planet=257B
[2026-06-29 18:48:27] Setting up network namespaces and bridge
[2026-06-29 18:48:27] Setting up netns mount point...
[2026-06-29 18:48:27] Cleaning up existing network resources...
[2026-06-29 18:48:27] Creating network namespaces...
[2026-06-29 18:48:27] Network namespaces created successfully
[2026-06-29 18:48:27] Creating bridge br0...
[2026-06-29 18:48:27] Configuring ns-root...
[2026-06-29 18:48:27] Configuring ns-controller...
[2026-06-29 18:48:27] Configuring ns-node1...
[2026-06-29 18:48:27] Configuring ns-node2...
[2026-06-29 18:48:27] Network namespaces configured:
[2026-06-29 18:48:27]   ns-root:       172.20.0.10:9993
[2026-06-29 18:48:27]   ns-controller: 172.20.0.11:9993
[2026-06-29 18:48:27]   ns-node1:      172.20.0.12:9993
[2026-06-29 18:48:27]   ns-node2:      172.20.0.13:9993
[2026-06-29 18:48:27]   Bridge:        172.20.0.1
[2026-06-29 18:48:27] Starting zerotier-one root on port 9993 in ns-root
[2026-06-29 18:48:27] zerotier-one root started with PID 104
[2026-06-29 18:48:27] Starting zerotier-one controller on port 9993 in ns-controller
[2026-06-29 18:48:27] zerotier-one controller started with PID 107
[2026-06-29 18:48:27] Starting zerotier-one node on port 9993 in ns-node1
[2026-06-29 18:48:27] zerotier-one node1 started with PID 111
[2026-06-29 18:48:27] Starting zerotier-one node on port 9993 in ns-node2
[2026-06-29 18:48:27] zerotier-one node2 started with PID 115
[2026-06-29 18:48:27] Waiting for zerotier-one-root on 172.20.0.10:9993...
[2026-06-29 18:48:28] zerotier-one-root is ready
[2026-06-29 18:48:28] Waiting for controller on 172.20.0.11:9993...
[2026-06-29 18:48:28] controller is ready
[2026-06-29 18:48:28] Waiting for node1 on 172.20.0.12:9993...
[2026-06-29 18:48:28] node1 is ready
[2026-06-29 18:48:28] Waiting for node2 on 172.20.0.13:9993...
[2026-06-29 18:48:28] node2 is ready
[2026-06-29 18:48:28] All services started successfully
[2026-06-29 18:48:28] Waiting for controller to fully initialize...
[2026-06-29 18:48:33] Waiting for authtoken.secret file to be created...
[2026-06-29 18:48:33] Creating network on controller
[2026-06-29 18:48:33] API Request: POST http://127.0.0.1:9993/controller/network (from ns-controller)
[2026-06-29 18:48:33] API Response Code: 200
[2026-06-29 18:48:33] API Response Body: {"authTokens":[null],"authorizationEndpoint":"","capabilities":[],"clientId":"","creationTime":1782758913775,"dns":[],"enableBroadcast":true,"id":"5946b73234c7bbe7","ipAssignmentPools":[],"mtu":2800,"multicastLimit":32,"name":"","nwid":"5946b73234c7bbe7","objtype":"network","private":true,"remoteTraceLevel":0,"remoteTraceTarget":null,"revision":1,"routes":[],"rules":[{"not":false,"or":false,"type":"ACTION_ACCEPT"}],"rulesSource":"","ssoConfig":{"ssoAuthorizationEndpoint":null,"ssoClientId":"","ssoIssuer":"","ssoLinkedId":"","ssoProvider":""},"ssoEnabled":false,"tags":[],"v4AssignMode":{"zt":false},"v6AssignMode":{"6plane":false,"rfc4193":false,"zt":false}}
[2026-06-29 18:48:33] Network created: 5946b73234c7bbe7
[2026-06-29 18:48:33] Configuring network settings
[2026-06-29 18:48:33] API Request: POST http://127.0.0.1:9993/controller/network/5946b73234c7bbe7 (from ns-controller)
[2026-06-29 18:48:33] API Response Code: 200
[2026-06-29 18:48:33] Network configured with IP pool 10.147.17.0/24
[2026-06-29 18:48:33] Getting node identities
[2026-06-29 18:48:33] Node 1 address: 3f033c9539
[2026-06-29 18:48:33] Node 2 address: 150817588f
[2026-06-29 18:48:33] Joining node 3f033c9539 to network 5946b73234c7bbe7
[2026-06-29 18:48:33] API Request: POST http://127.0.0.1:9993/network/5946b73234c7bbe7 (from ns-node1)
[2026-06-29 18:48:33] API Response Code: 200
[2026-06-29 18:48:33] Joining node 150817588f to network 5946b73234c7bbe7
[2026-06-29 18:48:33] API Request: POST http://127.0.0.1:9993/network/5946b73234c7bbe7 (from ns-node2)
[2026-06-29 18:48:33] API Response Code: 200
[2026-06-29 18:48:35] Authorizing nodes on controller
[2026-06-29 18:48:35] API Request: POST http://127.0.0.1:9993/controller/network/5946b73234c7bbe7/member/3f033c9539 (from ns-controller)
[2026-06-29 18:48:35] API Response Code: 200
[2026-06-29 18:48:35] API Request: POST http://127.0.0.1:9993/controller/network/5946b73234c7bbe7/member/150817588f (from ns-controller)
[2026-06-29 18:48:35] API Response Code: 200
[2026-06-29 18:48:35] Nodes authorized on network
[2026-06-29 18:48:37] Verifying network membership
[2026-06-29 18:48:37] Network 5946b73234c7bbe7 has 0 members
[2026-06-29 18:48:37] Dumping network configuration to network.json
[2026-06-29 18:48:37] Network configuration saved to /test-results/network.json
[2026-06-29 18:48:37] === ZeroTier Infrastructure Ready ===
[2026-06-29 18:48:37] 
[2026-06-29 18:48:37] Network Namespaces (bridge: 172.20.0.1/24, all on port 9993):
[2026-06-29 18:48:37]   ns-root:       172.20.0.10:9993 (74bea25db9)
[2026-06-29 18:48:37]   ns-controller: 172.20.0.11:9993 (5946b73234)
[2026-06-29 18:48:37]   ns-node1:      172.20.0.12:9993 (3f033c9539)
[2026-06-29 18:48:37]   ns-node2:      172.20.0.13:9993 (150817588f)
[2026-06-29 18:48:37] 
[2026-06-29 18:48:37] Network:        5946b73234c7bbe7
[2026-06-29 18:48:37] Logs:           /test-results/*.log
[2026-06-29 18:48:37] Network Config: /test-results/network.json
[2026-06-29 18:48:37] 
[2026-06-29 18:48:37] === ZeroTier-to-ZeroTier Connectivity Test ===
[2026-06-29 18:48:37] 
[2026-06-29 18:48:37] Waiting up to 60s for ZeroTier IP assignment...
[2026-06-29 18:48:48] Both nodes have an overlay IP after 10s
[2026-06-29 18:48:48] Node1 ZeroTier IP: 10.147.17.114
[2026-06-29 18:48:48] Node2 ZeroTier IP: 10.147.17.72
[2026-06-29 18:48:48] 
[2026-06-29 18:48:48] Testing ZeroTier connectivity (5 pings each direction)...
[2026-06-29 18:48:48] 
[2026-06-29 18:48:48] Node1 -> Node2 (10.147.17.114 -> 10.147.17.72):
[2026-06-29 18:48:52]   node1 -> node2 (10.147.17.72): OK
[2026-06-29 18:48:52] 
[2026-06-29 18:48:52] Node2 -> Node1 (10.147.17.72 -> 10.147.17.114):
[2026-06-29 18:48:56]   node2 -> node1 (10.147.17.114): OK
[2026-06-29 18:48:56] 
[2026-06-29 18:48:56] ZeroTier connectivity test: SUCCESS
[2026-06-29 18:48:56] 
[2026-06-29 18:48:56] === ZeroTier Overlay Throughput Test (iperf3, 10s) ===
[2026-06-29 18:48:56] 
[2026-06-29 18:48:56] Starting iperf3 server in ns-node2 (10.147.17.72:5201)
[2026-06-29 18:48:57] Running iperf3 node1 -> node2 (10.147.17.114 -> 10.147.17.72) for 10s...
Connecting to host 10.147.17.72, port 5201
[  5] local 10.147.17.114 port 48746 connected to 10.147.17.72 port 5201
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec   504 MBytes  4.22 Gbits/sec  162   1.05 MBytes       
[  5]   1.00-2.00   sec   519 MBytes  4.35 Gbits/sec    0   1.60 MBytes       
[  5]   2.00-3.00   sec   522 MBytes  4.38 Gbits/sec    0   2.01 MBytes       
[  5]   3.00-4.00   sec   522 MBytes  4.38 Gbits/sec    0   2.35 MBytes       
[  5]   4.00-5.00   sec   521 MBytes  4.37 Gbits/sec    1   1.86 MBytes       
[  5]   5.00-6.00   sec   524 MBytes  4.40 Gbits/sec    0   2.22 MBytes       
[  5]   6.00-7.00   sec   522 MBytes  4.38 Gbits/sec    0   2.52 MBytes       
[  5]   7.00-8.00   sec   521 MBytes  4.37 Gbits/sec   16   1.48 MBytes       
[  5]   8.00-9.00   sec   521 MBytes  4.37 Gbits/sec    0   1.91 MBytes       
[  5]   9.00-10.00  sec   521 MBytes  4.37 Gbits/sec  172   1.31 MBytes       
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec  5.08 GBytes  4.36 Gbits/sec  351             sender
[  5]   0.00-10.00  sec  5.07 GBytes  4.36 Gbits/sec                  receiver

iperf Done.
[2026-06-29 18:49:07] iperf3 overlay throughput test: SUCCESS
[2026-06-29 18:49:07] 
[2026-06-29 18:49:07] Writing result to /test-results/result.txt
[2026-06-29 18:49:07] Result: 1 (1=success, 0=failure)
[2026-06-29 18:49:07] 
[2026-06-29 18:49:07] === Node info ===
[2026-06-29 18:49:07] ns-root:
200 info 74bea25db9 1.16.2 ONLINE
[2026-06-29 18:49:07] ns-controller:
200 info 5946b73234 1.16.2 ONLINE
[2026-06-29 18:49:07] ns-node1:
200 info 3f033c9539 1.16.2 ONLINE
[2026-06-29 18:49:07] ns-node2:
200 info 150817588f 1.16.2 ONLINE
[2026-06-29 18:49:07] 
[2026-06-29 18:49:07] Snapshotting daemon home directories to /test-results/daemon-homes
[2026-06-29 18:49:07] Daemon home directories written to /test-results/daemon-homes
[2026-06-29 18:49:07] 
[2026-06-29 18:49:07] Complete
```
